File: src/org/cappuccino_project/ide/intellij/plugin/psi/interfaces/ObjJMethodHeaderDeclaration.java
package org.cappuccino_project.ide.intellij.plugin.psi.interfaces;

import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJMethodHeaderDeclarationStub;
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.MethodScope;
import org.jetbrains.annotations.NotNull;

public interface ObjJMethodHeaderDeclaration<StubT extends ObjJMethodHeaderDeclarationStub<? extends ObjJMethodHeaderDeclaration>> extends ObjJStubBasedElement<StubT>, ObjJCompositeElement, ObjJHasMethodSelector, ObjJHasContainingClass {

    @NotNull
    String getReturnType();

    @NotNull
    MethodScope getMethodScope();

    default boolean isStatic() {
        return getMethodScope() == MethodScope.STATIC;
    }

    StubT getStub();
}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/interfaces/ObjJMethodHeaderDeclaration.kt
package org.cappuccino_project.ide.intellij.plugin.psi.interfaces

import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJMethodHeaderDeclarationStub
import org.cappuccino_project.ide.intellij.plugin.psi.utils.ObjJMethodPsiUtils.MethodScope

interface ObjJMethodHeaderDeclaration<StubT : ObjJMethodHeaderDeclarationStub<out ObjJMethodHeaderDeclaration<*>>> : ObjJStubBasedElement<StubT>, ObjJCompositeElement, ObjJHasMethodSelector, ObjJHasContainingClass {

    val returnType: String

    val methodScope: MethodScope

    val isStatic: Boolean
        get() = methodScope == MethodScope.STATIC

    override fun getStub(): StubT
}
